hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/StateContext.java

```diff
@@ -332,19 +363,21 @@ public void putBackReports(List<GeneratedMessage> reportsToPutBack,
     return reportsToReturn;
   }
 
-  List<GeneratedMessage> getNonIncrementalReports() {
+  List<GeneratedMessage> getNonIncrementalReports(
+      InetSocketAddress endpoint) {
+    Map<String, AtomicBoolean> mp = fullReportSendIndicator.get(endpoint);
     List<GeneratedMessage> nonIncrementalReports = new LinkedList<>();
-    GeneratedMessage report = containerReports.get();
-    if (report != null) {
-      nonIncrementalReports.add(report);
-    }
-    report = nodeReport.get();
-    if (report != null) {
-      nonIncrementalReports.add(report);
-    }
-    report = pipelineReports.get();
-    if (report != null) {
-      nonIncrementalReports.add(report);
+    if (null != mp){
+      for (Map.Entry<String, AtomicBoolean> kv : mp.entrySet()) {
+        if (kv.getValue().get()) {
+          String reportType = kv.getKey();
+          GeneratedMessage msg = type2Reports.get(reportType).get();
+          if (null != msg) {
+            nonIncrementalReports.add(msg);
+            mp.get(reportType).set(false);
+          }
```

Comment:
IIUC `msg` here can never be null as long as the `reportType` is valid because of `initReportTypeCollection`.

Actually we might even want to throw exception if msg is indeed null somehow to catch potential problem earlier in the future.

Reply:
for now, maybe we can not do this change , because it will make several unit test fail.
see `TestDatanodeStateMachine#testDatanodeStateContext`, `TestStateContext#testReportAPIs`, and `TestHeartbeatEndpointTask#testheartbeatWithNodeReport`. in some test case , some of nonIncrementalReportType is not set, so `msg == null ` will be a normal behavior and the CI will fail with throwing an exception .  may be we can fix these case in another jira later.